torch/optim/adam.py

```diff
@@ -537,23 +566,37 @@ def _multi_tensor_adam(
                 )
 
         # Decay the first and second moment running average coefficient
-        torch._foreach_lerp_(device_exp_avgs, device_grads, 1 - beta1)
+        # Use device beta1 if beta1 is a tensor to ensure all
+        # tensors are on the same device
+        torch._foreach_lerp_(device_exp_avgs, device_grads, 1 - device_beta1)
 
         torch._foreach_mul_(device_exp_avg_sqs, beta2)
+
+        # Due to the strictness of the _foreach_addcmul API, we can't have a single
+        # tensor scalar as the scalar arg (only python number is supported there)
```

Comment:
@janeyx99 lmk what you think of this or if you have any better ideas here, but here's my reasoning for this. I can't use an inplace op here (we can't mutate device_grads) and I also can't pass a scalar tensor as the value argument (This is true for all of the foreach APIs, it can only be number or complex), so I had to add this extra out of place mul.

Reply:
Yea, I'm fine with this--could you also open an issue to track since the ideal would be that our foreach APIs can support Tensor values for addcmul

Reply:
Sounds good, I will file an issue.

Reply:
https://github.com/pytorch/pytorch/issues/139795